homeassistant/components/automation/__init__.py

```diff
@@ -210,6 +226,67 @@ async def reload_service_handler(service_call):
     return True
 
 
+class AutomationTrace:
+    """Container for automation trace."""
+
+    def __init__(self, entity_id, config, trigger):
+        """Container for automation trace."""
+        self._action_trace = None
+        self._condition_trace = None
+        self._config = config
+        self._error = None
+        self._entity_id = entity_id
+        self._timestamp_finish = None
+        self._timestamp_start = dt_util.utcnow()
+        self._trigger = trigger
+        self._variables = None
+
+    def set_error(self, ex):
+        """Set error."""
+        self._error = ex
+
+    def set_variables(self, variables):
+        """Set variables."""
+        self._variables = variables
+
+    def set_condition_trace(self, condition_trace):
+        """Set condition trace."""
+        self._condition_trace = condition_trace
+
+    def set_action_trace(self, action_trace):
+        """Set action trace."""
+        self._action_trace = action_trace
+
+    def finished(self):
+        """Set finish time."""
+        self._timestamp_finish = dt_util.utcnow()
+
+
+@contextmanager
+def trace_automation(hass, entity_id, config, trigger):
+    """Trace action execution."""
+    action_trace_clear()
+    automation_trace = AutomationTrace(entity_id, config, trigger)
+    hass.data[DATA_AUTOMATION_TRACE].append(automation_trace)
```

Comment:
Any reason you chose to store all traces in a single spot instead of attaching them to the automation entity ? The latter has the benefit (or con?) that reloading automations also reset it's traces.

Reply:
I was not sure how to store the data, but what's there now - one huge list - is probably not what we want.

If the traces are stored in the entities, we will lose all traces for all automations whenever automations are reloaded, that will be very annoying.

We should probably store the last `n` runs of each automation, potentially also with a cap of the total number of stored traces, but globally so they persist when automations are reloaded.

Reply:
That can be solved when a config hasn't changed, we shouldn't reload that automation entity and lose it's traces. (in a future PR).